### DISCUSSION TITLE
Template testing

### DISCUSSION THREAD

**Discussion by dkrupicka** [mojolicious/mojo#2115] [Q&A] [answered]
Hello,

is there a way how to test the code embedded in html templates, i.e. just pass the needed variables and test the result?

Best regards
David

**Comment by Akron** [comment, marked as answer]
Yes - and it's fairly simple. You can directly test templates using the `Mojo::Template` API like this:

```perl
#!/usr/bin/env perl
use Mojo::Template;
use Test::More;

my $mt = Mojo::Template->new;
my $template_rendering = $mt->vars(1)->render(<<'TEMP', {works => 'Also fun'});
<p><%= uc($works) %></p>
TEMP

is($template_rendering,"<p>ALSO FUN</p>\n");

done_testing;
```

So simply render the template with the passed values and check the result.

You can also check the code on a higher level (e.g. when not using `Mojo::Template` as your template engine) by using `Test::Mojo` like this:

```perl
#!/usr/bin/env perl
use Mojolicious::Lite;
use Test::Mojo;
use Test::More;

helper print_uc => sub {
  my ($c, $str) = @_;
  return uc($str);
};

get '/' => 'fun';

get '/fine' => sub {
  my $c = shift;
  $c->stash(works => 'FUN');
  $c->render('fun');
};

# Test the embedded code
my $t = Test::Mojo->new;

$t->get_ok('/')
  ->status_is(500)
  ;

$t->get_ok('/fine')
  ->status_is(200)
  ->text_is('p', 'FUN')
  ;

done_testing;

__DATA__
@@fun.html.ep
<p><%= print_uc($works) %></p>
```

**Reply by kraih** [reply]
Another option if you have a Mojolicious app too is to just make a fake controller instance and render the template that way (also avoiding full HTTP requests):
```perl
is $t->app->build_controller->render_to_string('fun', works => 'FUN'),    '...';
```

**Reply by dkrupicka** [reply]
Thank you both, that was exactly what I was looking for.

**Reply by dkrupicka** [reply]
I guess I will need some other tool then to traverse the html, or?

**Reply by dkrupicka** [reply]
What about the Test::XPath? But the version is 0.20...
Is there something better? I would choose XPath because it is fast. And we have a lot of templates, some with more than 5000 lines.

**Reply by s1037989** [reply]
To traverse the HTML, use Mojo::DOM -- there's nothing better!

**Reply by dkrupicka** [reply]
Thanks a lot, I will try!

**Reply by dkrupicka** [reply]
Although, citing the docu _It will even try to interpret broken HTML and XML, so you should not use it for validation._ ...
So maybe I could use e.g. HTML::Lint at first.

**Reply by ernstki** [reply]
> ```perl
> my $template_rendering = $mt->vars(1)->render(<<'TEMP', {works => 'Also fun'});
> ```

@Akron Do you personally have any intution for why `->vars(1)` should be required?

It's always the same as the number of keys passed in, at least in this simple example. So I just don't understand why I'd get

```console
Global symbol "$works" requires explicit package name
```

if I forgot the `->vars(1)` here.

**Reply by ernstki** [reply]
Oh, ha, it's somewhat clearer now that I see this in the docs

```
$mt      = $mt->vars($bool);
```

So it's not a number of variables, it's a true/false value. Total misunderstanding on my part. `*facepalm*`

I'm guessing this is because Perl [collapses sub-lists](https://perldoc.perl.org/perldata#List-value-constructors), and there'd be no way to tell whether a list or a hash of values were passed in otherwise.
